scheduler/src/cook/kubernetes/controller.clj

```diff
@@ -96,163 +117,258 @@
-                                 ; TODO: Implement :pod/unknown cases.
-                                 ; TODO: Implement :pod/pending cases.
-                                 ; TODO: Implement :pod/need-to-fail cases.
-                                 [:missing :pod/running] (kill-task api-client expected-state-dict pod)
-                                 [:missing :pod/pending] (kill-task api-client expected-state-dict pod)
-                                 [:missing :pod/need-to-fail] (kill-task api-client expected-state-dict pod)
-                                 [:missing :pod/succeeded] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
-                                 [:missing :pod/failed] (remove-finalization-if-set-and-delete api-client expected-state-dict pod)
-                                 [:missing :missing] nil ; this can come up due to the recur at the end
-                                 (do
-                                   (log/error "Unexpected state: "
-                                              (vector (or expected-state :missing) (or (:state synthesized-state) :missing))
-                                              "for pod" pod-name)
-                                   expected-state-dict))]
-      (when-not (expected-state-equivalent? expected-state-dict new-expected-state-dict)
-        (update-or-delete! expected-state-map pod-name new-expected-state-dict)
+      [pod-synthesized-state-modified (or (:state synthesized-state) :missing)
+       new-cook-expected-state-dict (case (or cook-expected-state :missing)
+                                      :cook-expected-state/completed
+                                      (case pod-synthesized-state-modified
+                                        :missing nil ; Cause this entry to be deleted by update-or-delete! called later down.
+                                        ; This is an exception. The writeback to datomic has occurred, so there's nothing to do except to delete the pod from kubernetes
+                                        ; and remove it from our tracking.
+                                        :pod/succeeded (delete-pod api-client pod)
+                                        :pod/failed (delete-pod api-client pod)
+                                        :pod/running (kill-pod-weird api-client cook-expected-state-dict k8s-actual-state-dict) ; Who resurrected this pod? Where did it come from? Do we have two instances of cook?
+                                        :pod/unknown (mark-pod-completed-and-kill-weird-task compute-cluster k8s-actual-state-dict pod) ; TODO: Should mark mea culpa retry
+                                        :pod/waiting (kill-pod-weird api-client cook-expected-state-dict k8s-actual-state-dict) ; Who resurrected this pod? Where did it come from? Do we have two instances of cook?
+                                        )
```

Comment:
This bracket can move to the previous line once we move the comments above. Same applies to all below.

Reply:
Ok